spring-integration-core/src/main/java/org/springframework/integration/store/AbstractMessageGroupStore.java

```diff
@@ -110,6 +108,16 @@ public void setLazyLoadMessageGroups(boolean lazyLoadMessageGroups) {
 
 	@Override
 	public void registerMessageGroupExpiryCallback(MessageGroupCallback callback) {
+		if (callback instanceof UniqueExpiryCallback) {
+			boolean uniqueExpiryCallbackPresent =
+					this.expiryCallbacks.stream()
+							.anyMatch(UniqueExpiryCallback.class::isInstance);
+
+			Assert.isTrue(!uniqueExpiryCallbackPresent,
+					"Only one instance of 'UniqueExpiryCallback' can be registered in the 'MessageGroupStore'." +
+							" Use separate 'MessageGroupStore' for each aggregator/resequencer.");
+		}
```

Comment:
Perhaps `logger.error()` in 5.1 and assert in 5.2?

Reply:
I agree about the `logger.error()`, because this exception is going to be some kind of breaking change in the point releases.
But at the same time I don't agree about `5.2`: in JIRA we talked about something like additional `owner` property on the `MessageGroup` to allow to share a `MessageStore`, but at the same time skip those groups from expiration in the aggregator which were not initiated by this particular bean.